### PR TITLE
chore: anti-flakiness spring offensive

### PR DESCRIPTION
**Pull request by dpc**
This has been on my mind for a long while: Run the tests in the merge queue multiple times to make detecting flakiness easier.

Currently there's some issue surfaced with #4250 , so *temporarily* disable random peer failures.

### CODE REVIEW

File: nix/flakebox.nix [hunk 1]
```diff
@@ -312,15 +312,33 @@ rec {
...
+    buildPhaseCargoCommand = "source <(cargo llvm-cov show-env --export-prefix); cargo build --locked --workspace --all-targets --profile $CARGO_PROFILE;";
+    nativeBuildInputs = [ pkgs.cargo-llvm-cov ];
+    doCheck = false;
+  };
+
+  workspaceTestCovBase = { times }: craneLib.buildPackage {
+    pname = "fedimint-workspace-lcov";
+    cargoArtifacts = workspaceCov;
+    buildPhaseCargoCommand = lib.concatStringsSep "\n" (
+      lib.replicate times ''
```
Comment (bradleystachurski) [inline nix/flakebox.nix:327]: Neat!

File: fedimint-server/src/net/connect.rs [hunk 2]
```diff
@@ -576,7 +576,7 @@ pub mod mock {
 
     impl StreamReliability {
         pub const MILDLY_UNRELIABLE: StreamReliability = {
-            let failure_rate = FailureRate(0.1);
+            let failure_rate = FailureRate(0.0);
```
Comment (bradleystachurski) [inline fedimint-server/src/net/connect.rs:579]: Should we make an issue to make sure we re-enable random peer failures?

**Review by bradleystachurski — APPROVED** [review]
I like the idea 👍 

Excited to see if it catches some flakes.

**Comment by bradleystachurski** [comment]
Might be worth surfacing: I'd anticipate this will add ~30 minutes in CI for merges

- A quick scan of recent test steps ([example](https://github.com/fedimint/fedimint/actions/runs/7819460662/job/21331947966)) take ~8 minutes
- Running 5 times in merge queue => 40 minutes, thus ~30 additional

I'd assume you're implicitly okay with that, just want to surface.

**Comment by justinmoon** [comment]
One thing I've been wondering about -- would there be any way to automatically record when a specific test fails and then passes when manually retried (without code changing at all)? Basically automate what we do manually in https://github.com/fedimint/fedimint/issues/2825.

**Comment by dpc** [comment]
> One thing I've been wondering about -- would there be any way to automatically record when a specific test flakes? 

The hope is that it won't be necessary because everything will keep failing all the time.

File: nix/flakebox.nix [hunk 3]
```diff
@@ -357,24 +375,29 @@ rec {
...
+        patchShebangs ./scripts
+        export FM_CARGO_DENY_COMPILATION=1
+        ./scripts/tests/test-ci-all.sh || exit 1
+        sed -i -e 's/exit 0/exit 1/g' scripts/tests/always-success-test.sh
+        echo "Verifying failure detection..."
+        ./scripts/tests/test-ci-all.sh 1>/dev/null 2>/dev/null && exit 1
+      ''
+    );
```
Comment (maan2003) [inline nix/flakebox.nix:395]: needs more parallelism? maybe do this in ci-test-all.sh

Comment (dpc) [reply nix/flakebox.nix:395]: I don't know if it does. At some point we'll start hitting memory limits etc. We're already waiting so long for backward compt tests in Merge Queue and nobody seems to complain, so I guess this should be fine for now too.

**Review by maan2003 — DISMISSED** [review]

**Review by maan2003 — APPROVED** [review]

**Comment by maan2003** [comment]
lets see if it passed the merge queue

**Comment by maan2003** [comment]
is this "merge-base changed" bugged? I see no changes

**Review by maan2003 — DISMISSED** [review]

**Review by bradleystachurski — DISMISSED** [review]
> is this "merge-base changed" bugged? I see no changes

I believe it's because there's a shared commit that was merged from https://github.com/fedimint/fedimint/pull/4277

**Review by maan2003 — DISMISSED** [review]

**Comment by bradleystachurski** [comment]
This does seem quite buggy now. @dpc think it's worth rebasing so https://github.com/fedimint/fedimint/commit/25acaa101a744509a1660aa3893c51eec8ae2190 is no longer a part of this PR?

**Review by bradleystachurski — DISMISSED** [review]

**Review by maan2003 — APPROVED** [review]

**Comment by dpc** [comment]
Rebased.

As for time it takes - easy to lower.

**Comment by maan2003** [comment]
might be interesting to mark some tests as flaky and merge this PR, so we don't introduce any new flakes
https://nexte.st/book/retries.html

**Comment by dpc** [comment]
> might be interesting to mark some tests as flaky and merge this PR, so we don't introduce any new flakes https://nexte.st/book/retries.html

Oh. Never used that. Yeah, we can do that. But in practice it will mean giving up on fixing the problem, as no one is paying attention to test outputs in the CI. Better than nothing though.

**Review by bradleystachurski — DISMISSED** [review]

**Review by maan2003 — DISMISSED** [review]

**Comment by maan2003** [comment]
> fedimint-workspace-lcov-ci> 2024-02-09T22:55:38.328326Z  WARN new:task{name="listen task"}: net::peer: Error while opening incoming connection mint=PeerId(3) err=Randomly failed

didn't we remove random failures

**Comment by dpc** [comment]
```
fedimint-test-all-ci> parallel: This should not happen. You have found a bug. Please follow
fedimint-test-all-ci> https://www.gnu.org/software/parallel/man.html#reporting-bugs
fedimint-test-all-ci> Include this in the report:
fedimint-test-all-ci> * The version number: 20230922
fedimint-test-all-ci> * The bugid: Less than 1 byte memory free
fedimint-test-all-ci> * The command line being run
fedimint-test-all-ci> * The files being read (put the files on a webserver if they are big)
fedimint-test-all-ci> If you get the error on smaller/fewer files, please include those instead.
```

I wanted bugs in Fedimint... not in GNU Parallel...

**Comment by dpc** [comment]
> didn't we remove random failures

I think there might be two types of random failures and one is one tests, on the others.

**Comment by dpc** [comment]
Pushed a fix: we didn't run tests on macos, so we shouldn't run 5 tests either.

**Review by bradleystachurski — APPROVED** [review]

**Review by maan2003 — DISMISSED** [review]

**Review by justinmoon — DISMISSED** [review]

**Review by bradleystachurski — DISMISSED** [review]

File: scripts/tests/test-ci-all.sh [hunk 4]
```diff
@@ -33,6 +33,11 @@ cargo nextest run --no-run ${CARGO_PROFILE:+--cargo-profile ${CARGO_PROFILE}} ${
 # If you really need to break this rule, ping dpc
 export FM_CARGO_DENY_COMPILATION=1
 
+function rust_unit_tests() {
```
Comment (bradleystachurski) [inline scripts/tests/test-ci-all.sh:36]: Whoa, have we not been running unit tests in CI?

File: nix/flakebox.nix [hunk 5]
```diff
@@ -320,15 +320,31 @@ rec {
...
+    ''
+    );
     installPhaseCommand = "true";
     nativeBuildInputs = [ pkgs.cargo-llvm-cov ];
     doCheck = false;
   };
 
+  workspaceTestCov = workspaceTestCovBase { times = 1; };
+  workspaceTest5TimesCov = workspaceTestCovBase { times = 5; };
+  workspaceTest10TimesCov = workspaceTestCovBase { times = 10; };
```
Comment (bradleystachurski) [inline nix/flakebox.nix:346]: [nit] unused

File: gateway/ln-gateway/src/lib.rs [hunk 6]
```diff
@@ -257,6 +257,11 @@ pub struct Gateway {
     // incoming HTLCs.
     clients: FederationToClientMap,
 
+    /// Joining new Federation is protected by this lock to prevent trying to
```
Comment (bradleystachurski) [inline gateway/ln-gateway/src/lib.rs:260]: Would you prefer to first land/rebase https://github.com/fedimint/fedimint/pull/4295?

**Review by bradleystachurski — DISMISSED** [review]

File: gateway/ln-gateway/src/lib.rs (hunk 6 above)
Comment (bradleystachurski) [reply gateway/ln-gateway/src/lib.rs:260]: Same question for https://github.com/fedimint/fedimint/pull/4294

File: scripts/tests/test-ci-all.sh (hunk 4 above)
Comment (dpc) [reply scripts/tests/test-ci-all.sh:36]: They are being run in the Code Coverage step. But now that I'd like to run them multiple times it makes sense to run them here in parallel with everything as well.

File: gateway/ln-gateway/src/lib.rs (hunk 6 above)
Comment (dpc) [reply gateway/ln-gateway/src/lib.rs:260]: We should review and land in parallel. They are the exact same change so there's not conflict. And this change might take a while as I keep discovering things to fix.

File: scripts/tests/test-ci-all.sh (hunk 4 above)
Comment (elsirion) [reply scripts/tests/test-ci-all.sh:36]: After this commit, should we undo running coverage tests 5 times? Seems like this should have it covered.

**Review by elsirion — APPROVED** [review]

**Comment by dpc** [comment]
> After this commit, should we undo running coverage tests 5 times? Seems like this should have it covered.

There might be some differences in the run time performance due to ccov overhead, so maybe if could expose some stuff that doesn't show up here, but sure, we can tweak all we want.

**Comment by elsirion** [comment]
Aaaaaaand it flaked again :see_no_evil:

**Comment by dpc** [comment]
> Aaaaaaand it flaked again 🙈

I think it was one of the fixes I already had but didn't push with it yet. Let's try again.

File: gateway/ln-gateway/src/lib.rs (hunk 6 above)
Comment (dpc) [reply gateway/ln-gateway/src/lib.rs:260]: always be landing, whatever you can, whenever you can. :)

**Review by elsirion — APPROVED** [review]

**Review by maan2003 — DISMISSED** [review]

**Review by justinmoon — APPROVED** [review]

**Review by bradleystachurski — APPROVED** [review]

**Pull request merged**